Adopt more smart pointers in RemoteImageDecoderAVFProxy and RemoteLegacyCDMSessionProxy

RupinMittal opened the pull request:
#### e9e8ec4d1acd086c42721398e553c170eea92736
<pre>
Adopt more smart pointers in RemoteImageDecoderAVFProxy and RemoteLegacyCDMSessionProxy
<a href="https://bugs.webkit.org/show_bug.cgi?id=279878">https://bugs.webkit.org/show_bug.cgi?id=279878</a>
<a href="https://rdar.apple.com/136209504">rdar://136209504</a>

Reviewed by Youenn Fablet.

Smart pointer adoption as per the static analyzer.

* Source/WebKit/GPUProcess/media/RemoteImageDecoderAVFProxy.cpp:
(WebKit::RemoteImageDecoderAVFProxy::deleteDecoder):
(WebKit::RemoteImageDecoderAVFProxy::encodedDataStatusChanged):
(WebKit::RemoteImageDecoderAVFProxy::setExpectedContentSize):
(WebKit::RemoteImageDecoderAVFProxy::createFrameImageAtIndex):
(WebKit::RemoteImageDecoderAVFProxy::clearFrameBufferCache):
* Source/WebKit/GPUProcess/media/RemoteLegacyCDMSessionProxy.cpp:
(WebKit::RemoteLegacyCDMSessionProxy::sendMessage):
(WebKit::RemoteLegacyCDMSessionProxy::sendError):

Canonical link: <a href="https://commits.webkit.org/283891@main">https://commits.webkit.org/283891@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/49e719b983e4c51e21438288ed2ddb041e89b283

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67662 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47044 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20301 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71774 "Built successfully") | [⏳ 🛠 wincairo ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54843 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18606 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 wincairo-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70729 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43155 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58513 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/34654 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39829 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17161 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61792 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16260 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73414 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11625 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15569 "Found 1 new test failure: fast/loader/subframe-navigate-during-main-frame-load.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61614 "Found 1 new test failure: http/tests/security/contentSecurityPolicy/block-all-mixed-content/insecure-iframe-in-main-frame.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11660 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58581 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/61663 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9487 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3102 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10291 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/42847 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43925 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45114 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43666 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/GPUProcess/media/RemoteImageDecoderAVFProxy.cpp:
```diff
@@ -37,6 +37,8 @@
 #include <WebCore/GraphicsContext.h>
 #include <WebCore/ImageDecoderAVFObjC.h>
+#include <wtf/Ref.h>
+#include <wtf/RefPtr.h>
```
RefPtr.h is probably including Ref.h

RupinMittal replied:
Fixed!

youennf commented on Source/WebKit/GPUProcess/media/RemoteImageDecoderAVFProxy.cpp:
```diff
@@ -88,14 +90,14 @@ void RemoteImageDecoderAVFProxy::deleteDecoder(ImageDecoderIdentifier identifier
 
-    auto imageDecoder = m_imageDecoders.get(identifier);
-    connection->connection().send(Messages::RemoteImageDecoderAVFManager::EncodedDataStatusChanged(identifier, imageDecoder->frameCount(), imageDecoder->size(), imageDecoder->hasTrack()), 0);
+    RefPtr imageDecoder = m_imageDecoders.get(identifier);
```
Instead of doing contains and get, can we just do get and then `if (!imageDecoder) return;`?

RupinMittal replied:
That is much cleaner--Fixed!

youennf commented on Source/WebKit/GPUProcess/media/RemoteImageDecoderAVFProxy.cpp:
```diff
@@ -104,7 +106,7 @@ void RemoteImageDecoderAVFProxy::setExpectedContentSize(ImageDecoderIdentifier i
         return;
 
-    m_imageDecoders.get(identifier)->setExpectedContentSize(expectedContentSize);
+    RefPtr { m_imageDecoders.get(identifier) }->setExpectedContentSize(expectedContentSize);
```
I wonder whether we should try having HashMap::get() return a RefPtr instead of a raw pointer.
Not in this patch though.

RupinMittal replied:
Oh interesting. I didn't realize that HashMap::get() doesn't return the exact type of the values.

youennf commented on Source/WebKit/GPUProcess/media/RemoteImageDecoderAVFProxy.cpp:
```diff
@@ -140,7 +142,7 @@ void RemoteImageDecoderAVFProxy::createFrameImageAtIndex(ImageDecoderIdentifier
         return;
 
-    auto nativeImage = NativeImage::createTransient(m_imageDecoders.get(identifier)->createFrameImageAtIndex(index));
+    auto nativeImage = NativeImage::createTransient(RefPtr { m_imageDecoders.get(identifier) }->createFrameImageAtIndex(index));
```
Ditto here, I would write it:
```
RefPtr imageDecoder = m_imageDecoders.get(identifier);
if (!imageDecoder)
    return;
```

RupinMittal replied:
Fixed!

youennf commented on Source/WebKit/GPUProcess/media/RemoteLegacyCDMSessionProxy.cpp:
```diff
@@ -37,6 +37,7 @@
 #include <WebCore/LegacyCDM.h>
 #include <WebCore/SharedBuffer.h>
 #include <wtf/LoggerHelper.h>
+#include <wtf/Ref.h>
```
Is this one needed? I would have guessed that it was already included by now.

RupinMittal replied:
You're right, it's not needed--Fixed!

RupinMittal committed:
Adopt more smart pointers in RemoteImageDecoderAVFProxy and RemoteLegacyCDMSessionProxy
https://bugs.webkit.org/show_bug.cgi?id=279878
rdar://136209504

Reviewed by Youenn Fablet.

Smart pointer adoption as per the static analyzer.

* Source/WebKit/GPUProcess/media/RemoteImageDecoderAVFProxy.cpp:
(WebKit::RemoteImageDecoderAVFProxy::deleteDecoder):
(WebKit::RemoteImageDecoderAVFProxy::encodedDataStatusChanged):
(WebKit::RemoteImageDecoderAVFProxy::setExpectedContentSize):
(WebKit::RemoteImageDecoderAVFProxy::createFrameImageAtIndex):
(WebKit::RemoteImageDecoderAVFProxy::clearFrameBufferCache):
* Source/WebKit/GPUProcess/media/RemoteLegacyCDMSessionProxy.cpp:
(WebKit::RemoteLegacyCDMSessionProxy::sendMessage):
(WebKit::RemoteLegacyCDMSessionProxy::sendError):

Canonical link: https://commits.webkit.org/283891@main
